tests/parametric/test_otel_span_methods.py

```diff
@@ -396,6 +398,212 @@ def test_otel_set_attributes_separately(self, test_agent, test_library):
+    @missing_feature(context.library == "python", reason="Not implemented")
+    @missing_feature(context.library == "python_http", reason="Not implemented")
+    def test_otel_span_started_with_link_from_another_span(self, test_agent, test_library):
+        """Test adding a span link created from another span.
+        This tests the functionality of "create a direct link between two spans
+        given two valid span (or SpanContext) objects" as specified in the RFC.
+        """
+        with test_library:
+            with test_library.otel_start_span("root") as parent:
+                parent.end_span()
+                with test_library.otel_start_span(
+                    "child",
+                    parent_id=parent.span_id,
+                    links=[Link(parent_id=parent.span_id, attributes={"foo": "bar", "array": ["a", "b", "c"]})],
+                ) as child:
+                    child.end_span()
+
+        traces = test_agent.wait_for_num_traces(1)
+        trace = find_trace_by_root(traces, otel_span(name="root"))
+        assert len(trace) == 2
+
+        root = find_span(trace, otel_span(name="root"))
+        child = find_span(trace, otel_span(name="child"))
+        assert child.get("parent_id") == root.get("span_id")
+
+        span_links = retrieve_span_links(child)
+        assert span_links is not None
+        assert len(span_links) == 1
+
+        link = span_links[0]
+        assert link.get("span_id") == root.get("span_id")
+        assert link.get("trace_id") == root.get("trace_id")
+        root_tid = root["meta"].get("_dd.p.tid") or "0" if "meta" in root else "0"
+        assert (link.get("trace_id_high") or 0) == int(root_tid, 16)
+        assert link["attributes"].get("foo") == "bar"
+        assert link["attributes"].get("array.0") == "a"
+        assert link["attributes"].get("array.1") == "b"
+        assert link["attributes"].get("array.2") == "c"
+        # TODO Why does flags is expected to be 0, meaning "not set"? 
+        # If 0, it is not sampled so it won't be send to the agent.
+        # assert (link.get("flags") or 0) == 0
+
+    @missing_feature(context.library < "java@1.26.0", reason="Implemented in 1.26.0")
+    @missing_feature(context.library == "nodejs", reason="Not implemented")
+    @missing_feature(context.library == "dotnet", reason="Not implemented")
+    @missing_feature(context.library == "python", reason="Not implemented")
+    @missing_feature(context.library == "python_http", reason="Not implemented")
+    def test_otel_span_started_with_link_from_datadog_headers(self, test_agent, test_library):
+        """Properly inject datadog distributed tracing information into span links.
+        """
+        with test_library:
+            with test_library.otel_start_span(
+                "root",
+                links=[
+                    Link(
+                        http_headers=[
+                            ["x-datadog-trace-id", "1234567890"],
```

Comment:
It is the purpose of the test: to create span links from `distributed tracing information`, so headers here.
The parametric client app is supposed to treat differently programmatic link creation (the thing you suggest) from link creation of extracted tracing information.
This test triggers the second case.

Reply:
According to the spec if there is an active span in a process (Span A) and then ddtrace distributed headers are received (contain data about Span B). Then Span A should be the **parent** of nested spans (Span C) and Span B should be a **link**. To test this logic we do not need to send a link object to the parametric app. We just need to:
- Create a span (Span A)
- Send distributed tracing headers (containing info about Span B)
- Create a second Span (Span C)
- End Span A and Span C
- Then ensure: Span A <- parent-child -> Span C, Span B <- linked -> Span C

I am not sure if I am missing something here

Reply:
The test is here not related to the case you describe.
It is use to test the span link creation from distributed tracing informations:

```
There must be functionality to:
- create a span link from distributed tracing information (e.g. datadog, b3, w3c headers)
```

Reply:
The specification has `what need to be tested` but it appears to be missing `how` distributed tracing information should be used to generate span links. The RFC makes some recommendations but it's missing definitive statements. The example I gave above is the most concrete [example](https://docs.google.com/document/d/1RcTjO8ER5FNweJ0rqVcQEbf6tb1PrlPsQ6coZZww8Xg/edit#heading=h.f7rbvhg00e8f) I could find but you're right it may not be relevant to this existing test.

I have these two high level questions:
1. Should there be a special value in the tracestate/x-datadog-tags to distinguish between when span links or parent-child relationships could be created? [decision made by the client/producer]
2. Should the downstream application (ex: consumer/server) make the decision to generate a link depending on the context [multiple parents/sources, special integration is used, traced servers, etc]? 
   - I think this is what we are trying to do but it would be nice if we could create an exhaustive list and have a test case for each scenario.

 
@bwoebi What are your thoughts here? Should we spell out some specific test cases in the RFC?

Reply:
The python tracer can generate span links from distributed tracing using the following code:

```python
 # http_headers={"x-datadog-trace-id": "1234567890", "x-datadog-parent-id": "9876543210", "x-datadog-sampling-priority": "2", "x-datadog-tags": "_dd.p.dm=-4,_dd.p.tid=0000000000000010"}

from ddtrace import tracer
from ddtrace.propagation.http import HTTPPropagator

context = HTTPPropagator.extract(headers=http_headers)
span = tracer.trace("span")
span.link_span(context)
```

3. Is this sufficient to say the python tracer supports generating span links from distributed tracing headers or should there be a helper/api that encapsulate this logic?

Reply:
1. I don't really understand which distinction you mean? You mean providing information in producing contexts as part of the distributed headers, so that consumers may decide to create new linked traces or do a distributed trace? This isn't really part of the specification, because I haven't seen the need for it. There's no fundamental reason not to have it though. What we have been doing in PHP instead is having the caller create a new linked trace - and propagate this spans information to downstream.
2. By default a downstream application should make the decision to do a distributed trace, for now at least. If you have multiple parents/sources, span links are the only realistic option. The RFC talks about this, but ultimately, these may need to be judgment calls what's easiest to use. Eventually we'll move to span links first, when we get better UI and UX for span links.
3. I think this is fully sufficient.

Reply:
Awesome @bwoebi. This really clears things up for me. Thanks for the detailed response. I don't think we need to make any changes here 😄